### PR TITLE
HTBHF-632 Added error handler library.

### PR DESCRIPTION
**Pull request by dwybourn**

### CODE REVIEW

File: error_handler/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Request Context Module
```
Comment (YetAnotherMatt) [inline error_handler/README.md]: This readme needs updating

Comment (dwybourn) [reply error_handler/README.md]: Done.

File: error_handler/build.gradle [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    implementation "org.springframework.boot:spring-boot-autoconfigure:${springBootVersion}"
+    implementation "org.springframework:spring-webmvc:${springVersion}"
+    implementation "org.springframework:spring-context:${springVersion}"
+    implementation "com.fasterxml.jackson.datatype:jackson-datatype-jsr310:${jacksonVersion}"
+    implementation "com.google.code.findbugs:annotations:3.0.1"
+    implementation "org.slf4j:slf4j-api:1.7.25"
+    implementation "org.apache.tomcat.embed:tomcat-embed-core:9.0.12"
+    implementation "io.springfox:springfox-swagger2:${springFoxVersion}"
+    implementation "io.springfox:springfox-bean-validators:${springFoxVersion}"
+    implementation "uk.gov.dhsc.htbhf:htbhf-request-context:0.0.2"
```
Comment (YetAnotherMatt) [inline error_handler/build.gradle]: Can we make this inherit from the project version? (Not sure if we can, might be worth a try)

**Review by YetAnotherMatt — DISMISSED** [review]

File: error_handler/README.md (hunk 1 above)
Comment (YetAnotherMatt) [reply error_handler/README.md]: The title still needs updating :)

File: error_handler/README.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+This sub-module contains all the code that is needed to return formatted error responses. 
+
+Adding `ErrorHandlerConfiguration` to your application context will provide you with the following bean:
+
+- `errorHnalder`: Controller advice that captures validation exceptions and returns an error response with field information, timestamp and request id. e.g.
```
Comment (gethyn1) [inline error_handler/README.md]: Typo `errorHnalder`

File: error_handler/build.gradle (hunk 2 above)
Comment (dwybourn) [reply error_handler/build.gradle]: We can't as that version won't exist in bintray until after the build.

File: error_handler/README.md (hunk 3 above)
Comment (dwybourn) [reply error_handler/README.md]: Fixed.

File: error_handler/README.md (hunk 1 above)
Comment (dwybourn) [reply error_handler/README.md]: Fixed.

File: error_handler/build.gradle [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    mavenCentral()
+    maven {
+        url  "https://dl.bintray.com/departmentofhealth-htbhf/maven"
+    }
+}
+
+
+dependencies {
+    implementation "org.springframework:spring-web:${springVersion}"
+    implementation "org.springframework.boot:spring-boot-autoconfigure:${springBootVersion}"
```
Comment (mikeSmith08) [inline error_handler/build.gradle]: I'm not keen on having anything Spring Boot in this common module, is this essential? In my opinion, common libraries should never control and mandate Spring Boot dependancies, it only leads to real problems in projects that use it.

File: error_handler/build.gradle (hunk 2 above)
Comment (mikeSmith08) [reply error_handler/build.gradle]: Can't we have a reference like in Maven to `${project.version}`, Gradle should have an equivalent? If not, we should be using Maven. :-) But seriously, this must be parameterised.

File: error_handler/build.gradle (hunk 4 above)
Comment (dwybourn) [reply error_handler/build.gradle]: This is needed for the ConditionalOnMissingBean, which could be taken out as there's only one bean in the configuration.

Comment (dwybourn) [reply error_handler/build.gradle]: Removed.

File: error_handler/build.gradle (hunk 2 above)
Comment (dwybourn) [reply error_handler/build.gradle]: It does but the build will fail as the request-context version won't exist anywhere until after the build.

File: common_rest/README.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# Error Handler Module
```
Comment (YetAnotherMatt) [inline common_rest/README.md]: Should we call it 'Common Rest Module' instead of 'Error Handler Module' here?

**Review by YetAnotherMatt — DISMISSED** [review]

File: common_rest/README.md (hunk 5 above)
Comment (dwybourn) [reply common_rest/README.md]: Updated title and description.

**Review by YetAnotherMatt — DISMISSED** [review]

File: common_rest/README.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
+# Common Rest Module
+
+This sub-module contains all the code that is needed to return formatted error responses and add the request id and session id to the MDC for logging purposes. 
+
+Adding `ErrorHandlerConfiguration` to your application context will provide you with the following bean:
```
Comment (mikeSmith08) [inline common_rest/README.md]: Should be `with the following beans:`.

File: common_rest/README.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+- `headerInterceptor`: Interceptor added to the provided `RestTemplate`
```
Comment (mikeSmith08) [inline common_rest/README.md]: Doesn't say what the interceptor actually does.

File: common_rest/src/test/java/uk/gov/dhsc/htbhf/errorhandler/ErrorHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+        // Then
+        assertThat(responseEntity).isNotNull();
+        assertThat(responseEntity.getBody()).isInstanceOf(ErrorResponse.class);
+        ErrorResponse errorResponse = (ErrorResponse) responseEntity.getBody();
+        assertThat(errorResponse.getRequestId()).isEqualTo(REQUEST_ID);
+        assertThat(errorResponse.getMessage()).isEqualTo(VALIDATION_ERROR_MESSAGE);
+        List<ErrorResponse.FieldError> fieldErrors = errorResponse.getFieldErrors();
+        assertThat(fieldErrors.size()).isEqualTo(2);
+        assertThat(fieldErrors.get(0).getField()).isEqualTo("field1");
```
Comment (mikeSmith08) [inline common_rest/src/test/java/uk/gov/dhsc/htbhf/errorhandler/ErrorHandlerTest.java]: I think these assertions could be refactored to remove duplication by having a method call `assertFieldErrorPresent` that looked for the presence of a FieldError (or field name and message) within the List as ordering isn't really important.

Comment (dwybourn) [reply common_rest/src/test/java/uk/gov/dhsc/htbhf/errorhandler/ErrorHandlerTest.java]: Created field error assertion method. Didn't create one using .contains() on the list as that requires additional code to construct the expected fieldError objects.

**Review by YetAnotherMatt — APPROVED** [review]

**Review by mikeSmith08 — APPROVED** [review]

**Pull request merged**